File: v12/interpreters/ts/src/interpreter/standard_errors.ts
import * as AST from "../ast";
import type { Interpreter } from "./index";
import type { RuntimeValue } from "./values";

export type StandardRuntimeErrorKind = "DivisionByZeroError" | "OverflowError" | "ShiftOutOfRangeError";

type StandardRuntimeErrorDetails = {
  operation?: string;
  shift?: bigint;
};

export class StandardRuntimeError extends Error {
  kind: StandardRuntimeErrorKind;
  operation?: string;
  shift?: bigint;

  constructor(kind: StandardRuntimeErrorKind, message: string, details: StandardRuntimeErrorDetails = {}) {
    super(message);
    this.kind = kind;
    this.operation = details.operation;
    this.shift = details.shift;
  }
}

const I32_MIN = -(1n << 31n);
const I32_MAX = (1n << 31n) - 1n;

function clampI32(value: bigint): bigint {
  if (value < I32_MIN || value > I32_MAX) return 0n;
  return value;
}

function resolveStandardErrorStruct(interp: Interpreter, name: string): AST.StructDefinition {
  const cached = interp.standardErrorStructs.get(name);
  if (cached) return cached;
  const candidates = [
    name,
    `able.core.errors.${name}`,
    `core.errors.${name}`,
    `errors.${name}`,
  ];
  for (const candidate of candidates) {
    try {
      const val = interp.globals.get(candidate);
      if (val && val.kind === "struct_def") {
        interp.standardErrorStructs.set(name, val.def);
        return val.def;
      }
    } catch {
      // ignore lookup errors
    }
  }
  for (const bucket of interp.packageRegistry.values()) {
    const val = bucket.get(name);
    if (val && val.kind === "struct_def") {
      interp.standardErrorStructs.set(name, val.def);
      return val.def;
    }
  }
  const placeholder = AST.structDefinition(name, [], "named");
  interp.standardErrorStructs.set(name, placeholder);
  return placeholder;
}

export function makeStandardErrorValue(
  interp: Interpreter,
  err: StandardRuntimeError,
): Extract<RuntimeValue, { kind: "error" }> {
  const def = resolveStandardErrorStruct(interp, err.kind);
  const entries: Array<[string, RuntimeValue]> = [];
  if (err.kind === "OverflowError") {
    const operation = err.operation ?? err.message;
    entries.push(["operation", { kind: "String", value: operation }]);
  }
  if (err.kind === "ShiftOutOfRangeError") {
    const shift = clampI32(err.shift ?? 0n);
    entries.push(["shift", { kind: "i32", value: shift }]);
  }
  const instance = interp.makeNamedStructInstance(def, entries) as Extract<RuntimeValue, { kind: "struct_instance" }>;
  return interp.makeRuntimeError(err.message, instance) as Extract<RuntimeValue, { kind: "error" }>;
}

File: v12/interpreters/go/pkg/interpreter/standard_errors.go
package interpreter

import (
	"errors"
	"fmt"
	"math/big"

	"able/interpreter-go/pkg/ast"
	"able/interpreter-go/pkg/runtime"
)

type standardRuntimeErrorKind string

const (
	standardDivisionByZero  standardRuntimeErrorKind = "DivisionByZeroError"
	standardOverflow        standardRuntimeErrorKind = "OverflowError"
	standardShiftOutOfRange standardRuntimeErrorKind = "ShiftOutOfRangeError"
)

type standardRuntimeError struct {
	kind      standardRuntimeErrorKind
	message   string
	operation string
	shift     int64
}

const (
	minI32 = int64(-1 << 31)
	maxI32 = int64(1<<31 - 1)
)

func (e standardRuntimeError) Error() string {
	return e.message
}

func newDivisionByZeroError() error {
	return standardRuntimeError{kind: standardDivisionByZero, message: "division by zero"}
}

func (i *Interpreter) StandardDivisionByZeroErrorValue() runtime.ErrorValue {
	return i.makeStandardErrorValue(standardRuntimeError{kind: standardDivisionByZero, message: "division by zero"})
}

func (i *Interpreter) StandardOverflowErrorValue(operation string) runtime.ErrorValue {
	message := operation
	if message == "" {
		message = "integer overflow"
	}
	return i.makeStandardErrorValue(standardRuntimeError{kind: standardOverflow, message: message, operation: operation})
}

func (i *Interpreter) StandardShiftOutOfRangeErrorValue(shift int64) runtime.ErrorValue {
	return i.makeStandardErrorValue(standardRuntimeError{kind: standardShiftOutOfRange, message: "shift out of range", shift: shift})
}

func newOverflowError(operation string) error {
	message := operation
	if message == "" {
		message = "integer overflow"
	}
	return standardRuntimeError{kind: standardOverflow, message: message, operation: operation}
}

func newShiftOutOfRangeError(shift int64) error {
	return standardRuntimeError{
		kind:    standardShiftOutOfRange,
		message: "shift out of range",
		shift:   shift,
	}
}

func (i *Interpreter) resolveStandardErrorStruct(name string) *runtime.StructDefinitionValue {
	if def, ok := i.standardErrorStructs[name]; ok && def != nil {
		return def
	}
	candidates := []string{
		name,
		"able.core.errors." + name,
		"core.errors." + name,
		"errors." + name,
	}
	for _, key := range candidates {
		if val, err := i.global.Get(key); err == nil {
			if def, conv := toStructDefinitionValue(val, key); conv == nil {
				i.standardErrorStructs[name] = def
				return def
			}
		}
	}
	for _, bucket := range i.packageRegistry {
		if val, ok := bucket[name]; ok {
			if def, conv := toStructDefinitionValue(val, name); conv == nil {
				i.standardErrorStructs[name] = def
				return def
			}
		}
	}
	placeholder := ast.NewStructDefinition(ast.NewIdentifier(name), nil, ast.StructKindNamed, nil, nil, false)
	def := &runtime.StructDefinitionValue{Node: placeholder}
	i.standardErrorStructs[name] = def
	return def
}

func (i *Interpreter) makeStandardErrorValue(err standardRuntimeError) runtime.ErrorValue {
	def := i.resolveStandardErrorStruct(string(err.kind))
	fields := map[string]runtime.Value{}
	switch err.kind {
	case standardOverflow:
		operation := err.operation
		if operation == "" {
			operation = err.message
		}
		fields["operation"] = runtime.StringValue{Val: operation}
	case standardShiftOutOfRange:
		shift := err.shift
		if shift < minI32 || shift > maxI32 {
			shift = 0
		}
		fields["shift"] = runtime.IntegerValue{Val: big.NewInt(shift), TypeSuffix: runtime.IntegerI32}
	}
	instance := &runtime.StructInstanceValue{
		Definition: def,
		Fields:     fields,
	}
	payload := map[string]runtime.Value{
		"value": instance,
	}
	return runtime.ErrorValue{
		Message: err.message,
		Payload: payload,
	}
}

func (i *Interpreter) makeIndexErrorValue(index int, length int) runtime.ErrorValue {
	def := i.resolveStandardErrorStruct("IndexError")
	fields := map[string]runtime.Value{
		"index": runtime.IntegerValue{
			Val:        big.NewInt(int64(index)),
			TypeSuffix: runtime.IntegerI64,
		},
		"length": runtime.IntegerValue{
			Val:        big.NewInt(int64(length)),
			TypeSuffix: runtime.IntegerI64,
		},
	}
	instance := &runtime.StructInstanceValue{
		Definition: def,
		Fields:     fields,
	}
	message := fmt.Sprintf("index %d out of bounds for length %d", index, length)
	return runtime.ErrorValue{
		Message: message,
		Payload: map[string]runtime.Value{"value": instance},
	}
}

func (i *Interpreter) wrapStandardRuntimeError(err error) error {
	if err == nil {
		return nil
	}
	if _, ok := err.(raiseSignal); ok {
		return err
	}
	var standardErr standardRuntimeError
	if ok := errors.As(err, &standardErr); ok {
		return raiseSignal{value: i.makeStandardErrorValue(standardErr)}
	}
	return err
}
